File: LeetCodeProblems/LeetCodeProblems-C++/LeetCodeProblems-C++/DynamicProgrammingProblems/ClimbStairs.cpp
//
//  ClimbStairs.cpp
//  LeetCodeProblems-C++
//
//  Created by Ficow on 2021/7/25.
//

#include "ClimbStairs.hpp"

int climbStairs(int n);

// https://leetcode-cn.com/problems/climbing-stairs
void ClimbStairs::run() {
    assert(climbStairs(10) == 89);
}









unordered_map<int, int> map;
int climbStairs(int n) {
    if (n < 2) return 1;
    if (map[n]) return map[n];
    int result = climbStairs(n - 1) + climbStairs(n - 2);
    map[n] = result;
    return result;
}


//int dpSolution1(int n) {
//    if (n < 2) return 1;
//    int a = 1, b = 1;
//    for (int i = 2; i <= n; i++) {
//        int t = a + b;
//        a = b;
//        b = t;
//    }
//    return b;
//}

File: LeetCodeProblems/LeetCodeProblems-Swift/LeetCodeProblems-Swift/DynamicProgrammingProblems/ClimbStairs.swift
//
//  ClimbStairs.swift
//  LeetCodeProblems-Swift
//
//  Created by Ficow on 2021/7/25.
//

import Foundation

// https://leetcode-cn.com/problems/climbing-stairs/#
final class ClimbStairs {
    func run() {
        printAndAssert(result: dpSolution1(2), expected: 2)
        printAndAssert(result: dpSolution1(3), expected: 3)
        printAndAssert(result: dpSolution1(6), expected: 13)
        printAndAssert(result: dpSolution1(10), expected: 89)

//        benchmark(identifier: "dpSolution1") {
//            _ = dpSolution1(90)
//        }
//        benchmark(identifier: "memoSolution1") {
//            _ = memoSolution1(90)
//        }
    }

    func dpSolution1(_ n: Int) -> Int {
        if n < 3 { return n }
        var a = 1, b = 1
        for _ in 2...n {
            (b, a) = (a + b, b)
        }
        return b
    }

    func memoSolution1(_ n: Int) -> Int {
        var memo = [Int: Int]()
        func fib(_ x: Int) -> Int {
            if x <= 1 { return 1 }
            if let y = memo[x] { return y }
            let result = fib(x - 1) + fib(x - 2)
            memo[x] = result
            return result
        }
        return fib(n)
    }
}
